### ISSUE TITLE
Variadic tuple assignment fail 

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#49138]
# Bug Report

### 🔎 Search Terms

variadic tuple bug 

### 🕗 Version & Regression Information

The same behavior is visible from 4.2.3 through 4.8dev
although it it as only been a bug since variadic tuples. 

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAEBUAsFMGdtLSB7ArgGwCagEYIPIDSiAlgHYDGCAbgIYBOJtmJFoALqgA7pygAU9OO1AVaZUAHNkocQE8A7jCEBKAFAhQ1aPVglkZWKAAsAOgBMpgMygAfidMAOUJmjVTagN5rQvjnK4EcABGc1AAXlAAbQAiAD0YgBpQU1T+WHZGMklQAB9cZGRecTzQVDJXADNyaEwVKIBdZJiAEhiGnz8KAwyOUNoALhCwyNiE5raGgG4-WdBNAGUUDGwyZBE8WQkdemR6Tt9uwxF2UJwh0Ijo+KSOelRoCfaZ+bAltCxQNY2EEp29tQAXzUQA)

### 💻 Code

```ts
// These should be OK since variadic tuples (rest can go anywhere)
// versions 4.2.3 ~ 4.8 dev.
{
    type T12 = ["^", ...(string | boolean | undefined)[], "$"]
    const t12a:T12 = ["^", "$"];        // Should not be an error
    const t12b:T12 = ["^", true, "$"];  // Should not be an error
}
```

### 🙁 Actual behavior

As shown. 

### 🙂 Expected behavior

Not errors, because a match exists.  

### Additional Info

I examined the code.  `checkArrayLiteral` in `checker.ts` performs a unidirection (left to right) search only,

However, the output from `createNormalizedTupleType` outputs a  normalized type that can have a single tuple in the middle (like the one in the example).

That requires a bidirectional search to get the right results.

Alternatively, `createNormalizedTupleType` could be forced to roll up everything after the first rest into a single rest.  But such an algorithm would result in more false postives (false passes) than necessary.  So a bidirectional search is preferable.

I can provide more details if necessary.  (I even have bi-directional search code that passes all the baselines tests).   Just ask if you want it.

**Comment by Andarist** [comment]
note from my issues scavenger hunt: this issue was extracted from https://github.com/microsoft/TypeScript/issues/49105 and the draft PR [here](https://github.com/microsoft/TypeScript/pull/49106) seems to implement a fix for this

**Comment by craigphicks** [comment]
Seems to have been fixed in 5.0.4

**Comment by Andarist** [comment]
This was fixed between 5.0.0-dev.20230218 and 5.0.0-dev.20230219, so likely in https://github.com/microsoft/TypeScript/pull/52769

**Comment by craigphicks** [comment]
@Andarist - Thank you very much for finding that!
